File: src/cherry_pick.zig
// SPDX-License-Identifier: MIT OR Apache-2.0
const std = @import("std");
const hash_mod = @import("hash.zig");
const object = @import("object.zig");
const store_mod = @import("store.zig");
const merge_mod = @import("merge.zig");
const th = @import("test_helpers.zig");
const Hash = hash_mod.Hash;
const Allocator = std.mem.Allocator;

pub const CherryPickResult = struct {
    tree_hash: Hash,
    conflicts: []merge_mod.Conflict,
    original_message: []const u8,
    allocator: Allocator,

    pub fn deinit(self: *CherryPickResult) void {
        self.allocator.free(self.original_message);
        for (self.conflicts) |c| {
            self.allocator.free(c.path);
        }
        if (self.conflicts.len > 0) {
            self.allocator.free(self.conflicts);
        }
    }

    pub fn hasConflicts(self: CherryPickResult) bool {
        return self.conflicts.len > 0;
    }
};

/// Cherry-pick a single commit onto the current tree using 3-way merge.
///
/// Algorithm:
/// 1. Load target commit, get its first parent hash (or null if root commit)
/// 2. Get parent's tree_hash (base) and target's tree_hash (theirs)
/// 3. Call mergeTrees(parent_tree, ours_tree, target_tree)
/// 4. Return result with merged tree hash + conflicts + original message
pub fn cherryPick(
    allocator: Allocator,
    store: *store_mod.ObjectStore,
    target_hash: Hash,
    ours_tree: Hash,
) !CherryPickResult {
    // Load the target commit
    var target_obj = try store.get(allocator, target_hash);
    defer target_obj.deinit(allocator);

    if (target_obj != .commit) return error.NotACommit;

    const target_commit = target_obj.commit;

    // Get parent's tree hash. If root commit (no parents), use null (empty tree).
    var parent_tree: ?Hash = null;
    if (target_commit.parents.len > 0) {
        var parent_obj = try store.get(allocator, target_commit.parents[0]);
        defer parent_obj.deinit(allocator);
        if (parent_obj != .commit) return error.NotACommit;
        parent_tree = parent_obj.commit.tree_hash;
    }

    // Dupe the message before we lose the reference (target_obj is deferred deinit)
    const original_message = try allocator.dupe(u8, target_commit.message);
    errdefer allocator.free(original_message);

    // 3-way merge: base=parent_tree, ours=ours_tree, theirs=target_tree
    var merge_result = try merge_mod.mergeTrees(
        allocator,
        store,
        parent_tree,
        ours_tree,
        target_commit.tree_hash,
    );
    // Take ownership of merge_result's conflicts; prevent merge deinit from freeing them.
    const tree_hash = merge_result.tree_hash;
    const conflicts = merge_result.conflicts;
    merge_result.conflicts = @constCast(&[_]merge_mod.Conflict{});
    merge_result.deinit();

    return CherryPickResult{
        .tree_hash = tree_hash,
        .conflicts = conflicts,
        .original_message = original_message,
        .allocator = allocator,
    };
}

// ===========================================================================
// Tests
// ===========================================================================

test "cherry-pick adds a file onto branch missing it" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var store = try store_mod.ObjectStore.init(tmp.dir);
    defer store.close();

    // Base: one file "a.txt"
    const blob_a = try th.makeBlob(allocator, &store, "aaa");
    const base_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_a },
    };
    const base_tree = try th.makeTree(allocator, &store, &base_entries);
    const base_commit = try th.makeCommit(allocator, &store, base_tree, &.{}, "initial");

    // Target commit: adds "b.txt" on top of base
    const blob_b = try th.makeBlob(allocator, &store, "bbb");
    const target_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_a },
        .{ .name = "b.txt", .mode = .blob, .object_hash = blob_b },
    };
    const target_tree = try th.makeTree(allocator, &store, &target_entries);
    const target_parents = [_]Hash{base_commit};
    const target_commit = try th.makeCommit(allocator, &store, target_tree, &target_parents, "add b.txt");

    // Our branch: same as base (only a.txt)
    // Cherry-pick target onto our tree
    var result = try cherryPick(allocator, &store, target_commit, base_tree);
    defer result.deinit();

    try std.testing.expect(!result.hasConflicts());
    try std.testing.expectEqualStrings("add b.txt", result.original_message);

    // Verify merged tree has both a.txt and b.txt
    var merged_obj = try store.get(allocator, result.tree_hash);
    defer merged_obj.deinit(allocator);
    try std.testing.expectEqual(@as(usize, 2), merged_obj.tree.entries.len);

    // Check entries
    var found_a = false;
    var found_b = false;
    for (merged_obj.tree.entries) |entry| {
        if (std.mem.eql(u8, entry.name, "a.txt")) {
            try std.testing.expectEqual(blob_a, entry.object_hash);
            found_a = true;
        } else if (std.mem.eql(u8, entry.name, "b.txt")) {
            try std.testing.expectEqual(blob_b, entry.object_hash);
            found_b = true;
        }
    }
    try std.testing.expect(found_a);
    try std.testing.expect(found_b);
}

test "cherry-pick with modify-modify conflict" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var store = try store_mod.ObjectStore.init(tmp.dir);
    defer store.close();

    // Base: "a.txt" with content "original"
    const blob_orig = try th.makeBlob(allocator, &store, "original");
    const base_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_orig },
    };
    const base_tree = try th.makeTree(allocator, &store, &base_entries);
    const base_commit = try th.makeCommit(allocator, &store, base_tree, &.{}, "initial");

    // Target commit: modifies "a.txt" to "theirs-change"
    const blob_theirs = try th.makeBlob(allocator, &store, "theirs-change");
    const target_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_theirs },
    };
    const target_tree = try th.makeTree(allocator, &store, &target_entries);
    const target_parents = [_]Hash{base_commit};
    const target_commit = try th.makeCommit(allocator, &store, target_tree, &target_parents, "change a.txt");

    // Our branch: also modified "a.txt" to "ours-change"
    const blob_ours = try th.makeBlob(allocator, &store, "ours-change");
    const ours_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_ours },
    };
    const ours_tree = try th.makeTree(allocator, &store, &ours_entries);

    // Cherry-pick: should produce a conflict
    var result = try cherryPick(allocator, &store, target_commit, ours_tree);
    defer result.deinit();

    try std.testing.expect(result.hasConflicts());
    try std.testing.expectEqual(@as(usize, 1), result.conflicts.len);
    try std.testing.expectEqualStrings("a.txt", result.conflicts[0].path);
    try std.testing.expectEqual(merge_mod.ConflictKind.modify_modify, result.conflicts[0].kind);
    try std.testing.expectEqualStrings("change a.txt", result.original_message);
}

test "cherry-pick a root commit (no parent)" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var store = try store_mod.ObjectStore.init(tmp.dir);
    defer store.close();

    // Root commit with "a.txt"
    const blob_a = try th.makeBlob(allocator, &store, "aaa");
    const root_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_a },
    };
    const root_tree = try th.makeTree(allocator, &store, &root_entries);
    const root_commit = try th.makeCommit(allocator, &store, root_tree, &.{}, "root commit");

    // Our tree has "b.txt" only
    const blob_b = try th.makeBlob(allocator, &store, "bbb");
    const ours_entries = [_]object.TreeEntry{
        .{ .name = "b.txt", .mode = .blob, .object_hash = blob_b },
    };
    const ours_tree = try th.makeTree(allocator, &store, &ours_entries);

    // Cherry-pick root commit (no parent -> null base tree)
    // merge(null, ours_tree, root_tree) => should add a.txt from root, keep b.txt from ours
    var result = try cherryPick(allocator, &store, root_commit, ours_tree);
    defer result.deinit();

    try std.testing.expect(!result.hasConflicts());
    try std.testing.expectEqualStrings("root commit", result.original_message);

    // Verify merged tree has both files
    var merged_obj = try store.get(allocator, result.tree_hash);
    defer merged_obj.deinit(allocator);
    try std.testing.expectEqual(@as(usize, 2), merged_obj.tree.entries.len);
}

test "cherry-pick with delete-modify conflict" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var store = try store_mod.ObjectStore.init(tmp.dir);
    defer store.close();

    // Base: file "a.txt"
    const blob_a = try th.makeBlob(allocator, &store, "original");
    const base_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_a },
    };
    const base_tree = try th.makeTree(allocator, &store, &base_entries);
    const base_commit = try th.makeCommit(allocator, &store, base_tree, &.{}, "initial");

    // Target commit: removes "a.txt" (empty tree)
    const target_tree = try th.makeTree(allocator, &store, &.{});
    const target_parents = [_]Hash{base_commit};
    const target_commit = try th.makeCommit(allocator, &store, target_tree, &target_parents, "remove a.txt");

    // Ours: modified "a.txt"
    const blob_modified = try th.makeBlob(allocator, &store, "modified content");
    const ours_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_modified },
    };
    const ours_tree = try th.makeTree(allocator, &store, &ours_entries);

    // Cherry-pick: should produce a delete_modify conflict
    var result = try cherryPick(allocator, &store, target_commit, ours_tree);
    defer result.deinit();

    try std.testing.expect(result.hasConflicts());
    try std.testing.expectEqual(@as(usize, 1), result.conflicts.len);
    try std.testing.expectEqualStrings("a.txt", result.conflicts[0].path);
    try std.testing.expectEqual(merge_mod.ConflictKind.delete_modify, result.conflicts[0].kind);
}

test "cherry-pick adds multiple files" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var store = try store_mod.ObjectStore.init(tmp.dir);
    defer store.close();

    // Base: single file
    const blob_a = try th.makeBlob(allocator, &store, "aaa");
    const base_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_a },
    };
    const base_tree = try th.makeTree(allocator, &store, &base_entries);
    const base_commit = try th.makeCommit(allocator, &store, base_tree, &.{}, "initial");

    // Target commit adds 3 files on top of base
    const blob_b = try th.makeBlob(allocator, &store, "bbb");
    const blob_c = try th.makeBlob(allocator, &store, "ccc");
    const blob_d = try th.makeBlob(allocator, &store, "ddd");
    const target_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_a },
        .{ .name = "b.txt", .mode = .blob, .object_hash = blob_b },
        .{ .name = "c.txt", .mode = .blob, .object_hash = blob_c },
        .{ .name = "d.txt", .mode = .blob, .object_hash = blob_d },
    };
    const target_tree = try th.makeTree(allocator, &store, &target_entries);
    const target_parents = [_]Hash{base_commit};
    const target_commit = try th.makeCommit(allocator, &store, target_tree, &target_parents, "add b, c, d");

    // Ours is same as base
    var result = try cherryPick(allocator, &store, target_commit, base_tree);
    defer result.deinit();

    try std.testing.expect(!result.hasConflicts());

    // Verify merged tree has all 4 files
    var merged_obj = try store.get(allocator, result.tree_hash);
    defer merged_obj.deinit(allocator);
    try std.testing.expectEqual(@as(usize, 4), merged_obj.tree.entries.len);
}

test "cherry-pick non-commit returns error" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var store = try store_mod.ObjectStore.init(tmp.dir);
    defer store.close();

    // Store a blob and try to cherry-pick it
    const blob_hash = try th.makeBlob(allocator, &store, "just a blob");
    const empty_tree = try th.makeTree(allocator, &store, &.{});

    try std.testing.expectError(error.NotACommit, cherryPick(allocator, &store, blob_hash, empty_tree));
}

test "cherry-pick root commit with empty ours" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var store = try store_mod.ObjectStore.init(tmp.dir);
    defer store.close();

    // Root commit with "a.txt" and "b.txt"
    const blob_a = try th.makeBlob(allocator, &store, "aaa");
    const blob_b = try th.makeBlob(allocator, &store, "bbb");
    const root_entries = [_]object.TreeEntry{
        .{ .name = "a.txt", .mode = .blob, .object_hash = blob_a },
        .{ .name = "b.txt", .mode = .blob, .object_hash = blob_b },
    };
    const root_tree = try th.makeTree(allocator, &store, &root_entries);
    const root_commit = try th.makeCommit(allocator, &store, root_tree, &.{}, "root");

    // Ours is empty tree
    const empty_tree = try th.makeTree(allocator, &store, &.{});

    // Cherry-pick root onto empty: merge(null, empty, root_tree)
    // Both base and ours are empty, theirs has files -> should add them
    var result = try cherryPick(allocator, &store, root_commit, empty_tree);
    defer result.deinit();

    try std.testing.expect(!result.hasConflicts());

    // Result should have the target's tree content
    var merged_obj = try store.get(allocator, result.tree_hash);
    defer merged_obj.deinit(allocator);
    try std.testing.expectEqual(@as(usize, 2), merged_obj.tree.entries.len);
}

File: rust/crates/mkit-core/src/ops/cherry_pick.rs
//! Single-commit cherry-pick onto a different base tree.
//!
//! This is intentionally a *tree-level* operation: it computes the
//! 3-way merge of `(target.parents[0].tree, ours_tree, target.tree)`
//! and returns the resulting tree hash plus any conflicts. Building a
//! new commit on top of the merged tree is the caller's job — the CLI
//! layer wires refs and the index together.
//!
//! There is no `AlreadyAncestor` short-circuit here — that is a
//! higher-level decision the CLI makes before calling cherry-pick.

use crate::hash::Hash;
use crate::object::Object;
use crate::store::{ObjectStore, StoreError};

use super::merge::{self, Conflict};

/// Errors specific to cherry-pick on top of [`StoreError`]. We split
/// these out so callers can distinguish "your input hash didn't point
/// at a commit" (a programmer error) from filesystem failures.
#[derive(Debug, thiserror::Error)]
pub enum CherryPickError {
    #[error("target hash does not refer to a commit object")]
    NotACommit,
    #[error("target commit's first parent does not refer to a commit object")]
    ParentNotACommit,
    /// The target is a merge commit but no mainline parent was selected
    /// (git: "commit … is a merge but no -m option was given").
    #[error("commit is a merge but no mainline (-m <parent-number>) was given")]
    MergeNeedsMainline,
    /// A mainline parent was given for a non-merge commit (git: "mainline
    /// was specified but commit … is not a merge").
    #[error("mainline (-m) was specified but the commit is not a merge")]
    MainlineForNonMerge,
    /// The mainline parent number is out of range for this merge commit.
    #[error("invalid mainline parent {given}: commit has {parents} parents")]
    BadMainline { given: usize, parents: usize },
    #[error(transparent)]
    Store(#[from] StoreError),
}

/// Result of [`cherry_pick`]. `tree_hash` is the merged tree (always
/// written to the store, even on conflict — the merged tree contains
/// "ours" at every conflicting path). `original_message` is the target
/// commit's message verbatim, so the caller can use it as the basis
/// for a new commit.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CherryPickResult {
    pub tree_hash: Hash,
    pub conflicts: Vec<Conflict>,
    pub original_message: Vec<u8>,
}

impl CherryPickResult {
    #[must_use]
    pub fn has_conflicts(&self) -> bool {
        !self.conflicts.is_empty()
    }
}

/// Cherry-pick `target_hash` onto `ours_tree`.
///
/// Algorithm:
///
/// 1. Load the target commit. (Error if not a commit.)
/// 2. Load the target's first parent's tree as the merge `base`. If
///    the target is a root commit (no parents), `base = None` (empty
///    tree).
/// 3. 3-way merge `(base, ours_tree, target.tree_hash)`.
/// 4. Return the merged tree hash, any conflicts, and the target
///    commit's `message` so the caller can craft a new commit.
///
/// # Errors
///
/// * [`CherryPickError::NotACommit`] when `target_hash` doesn't point
///   at a commit object.
/// * [`CherryPickError::ParentNotACommit`] when the parent hash points
///   at something other than a commit.
/// * [`CherryPickError::Store`] for any wrapped store/serialize error.
pub fn cherry_pick(
    store: &ObjectStore,
    target_hash: Hash,
    ours_tree: Hash,
    mainline: Option<usize>,
) -> Result<CherryPickResult, CherryPickError> {
    let Object::Commit(target_commit) = store.read_object(&target_hash)? else {
        return Err(CherryPickError::NotACommit);
    };

    // Pick the parent whose diff is replayed, with git's `-m`/`--mainline`
    // semantics: a non-merge takes its sole parent (and rejects `-m`), while
    // a merge REQUIRES `-m <parent-number>` to choose the mainline (git
    // refuses to guess). The selected parent's tree is the base of the diff.
    let n_parents = target_commit.parents.len();
    let base_parent: Option<Hash> = match (n_parents, mainline) {
        (0, None) => None,
        (0 | 1, Some(_)) => return Err(CherryPickError::MainlineForNonMerge),
        (1, None) => Some(target_commit.parents[0]),
        (_, None) => return Err(CherryPickError::MergeNeedsMainline),
        (_, Some(m)) if m >= 1 && m <= n_parents => Some(target_commit.parents[m - 1]),
        (_, Some(m)) => {
            return Err(CherryPickError::BadMainline {
                given: m,
                parents: n_parents,
            });
        }
    };

    let parent_tree: Option<Hash> = match base_parent {
        None => None,
        Some(parent_hash) => {
            let Object::Commit(parent_commit) = store.read_object(&parent_hash)? else {
                return Err(CherryPickError::ParentNotACommit);
            };
            Some(parent_commit.tree_hash)
        }
    };

    let original_message = target_commit.message.clone();
    let merge_result = merge::merge_trees(
        store,
        parent_tree,
        Some(ours_tree),
        Some(target_commit.tree_hash),
    )?;

    Ok(CherryPickResult {
        tree_hash: merge_result.tree_hash,
        conflicts: merge_result.conflicts,
        original_message,
    })
}

// =====================================================================
// Tests
// =====================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use crate::object::{Blob, Commit, EntryMode, Identity, Object, Tree, TreeEntry};
    use crate::ops::merge::ConflictKind;
    use crate::serialize;
    use tempfile::TempDir;

    fn store() -> (TempDir, ObjectStore) {
        let d = TempDir::new().unwrap();
        let s = ObjectStore::init(d.path()).unwrap();
        (d, s)
    }
    fn put_blob(s: &ObjectStore, data: &[u8]) -> Hash {
        let bytes = serialize::serialize(&Object::Blob(Blob {
            data: data.to_vec(),
        }))
        .unwrap();
        s.write(&bytes).unwrap()
    }
    fn make_tree(s: &ObjectStore, entries: Vec<TreeEntry>) -> Hash {
        let bytes = serialize::serialize(&Object::Tree(Tree { entries })).unwrap();
        s.write(&bytes).unwrap()
    }
    fn entry(name: &[u8], mode: EntryMode, h: Hash) -> TreeEntry {
        TreeEntry {
            name: name.to_vec(),
            mode,
            object_hash: h,
        }
    }
    fn make_commit(s: &ObjectStore, tree: Hash, parents: &[Hash], message: &str) -> Hash {
        let c = Commit {
            tree_hash: tree,
            parents: parents.to_vec(),
            author: Identity::ed25519([0; 32]),
            signer: [0; 32],
            message: message.as_bytes().to_vec(),
            timestamp: message.len() as u64,
            message_hash: [0; 32],
            content_digest: [0; 32],
            signature: [0; 64],
        };
        s.write(&serialize::serialize(&Object::Commit(c)).unwrap())
            .unwrap()
    }
    fn tree_entries(s: &ObjectStore, h: Hash) -> Vec<TreeEntry> {
        match s.read_object(&h).unwrap() {
            Object::Tree(t) => t.entries,
            other => panic!("expected tree, got {other}"),
        }
    }

    #[test]
    fn adds_a_file_onto_branch_missing_it() {
        let (_d, s) = store();
        let blob_a = put_blob(&s, b"aaa");
        let base_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_a)]);
        let base_commit = make_commit(&s, base_tree, &[], "initial");
        let blob_b = put_blob(&s, b"bbb");
        let target_tree = make_tree(
            &s,
            vec![
                entry(b"a.txt", EntryMode::Blob, blob_a),
                entry(b"b.txt", EntryMode::Blob, blob_b),
            ],
        );
        let target_commit = make_commit(&s, target_tree, &[base_commit], "add b.txt");

        let r = cherry_pick(&s, target_commit, base_tree, None).unwrap();
        assert!(!r.has_conflicts());
        assert_eq!(r.original_message, b"add b.txt");
        let merged = tree_entries(&s, r.tree_hash);
        assert_eq!(merged.len(), 2);
    }

    #[test]
    fn modify_modify_conflict() {
        let (_d, s) = store();
        let blob_orig = put_blob(&s, b"original");
        let base_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_orig)]);
        let base_commit = make_commit(&s, base_tree, &[], "initial");
        let blob_theirs = put_blob(&s, b"theirs-change");
        let target_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_theirs)]);
        let target_commit = make_commit(&s, target_tree, &[base_commit], "change a.txt");
        let blob_ours = put_blob(&s, b"ours-change");
        let ours_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_ours)]);

        let r = cherry_pick(&s, target_commit, ours_tree, None).unwrap();
        assert!(r.has_conflicts());
        assert_eq!(r.conflicts.len(), 1);
        assert_eq!(r.conflicts[0].path, "a.txt");
        assert_eq!(r.conflicts[0].kind, ConflictKind::ModifyModify);
        assert_eq!(r.original_message, b"change a.txt");
    }

    #[test]
    fn root_commit_no_parent() {
        let (_d, s) = store();
        let blob_a = put_blob(&s, b"aaa");
        let root_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_a)]);
        let root_commit = make_commit(&s, root_tree, &[], "root commit");
        let blob_b = put_blob(&s, b"bbb");
        let ours_tree = make_tree(&s, vec![entry(b"b.txt", EntryMode::Blob, blob_b)]);
        let r = cherry_pick(&s, root_commit, ours_tree, None).unwrap();
        assert!(!r.has_conflicts());
        assert_eq!(r.original_message, b"root commit");
        assert_eq!(tree_entries(&s, r.tree_hash).len(), 2);
    }

    #[test]
    fn delete_modify_conflict() {
        let (_d, s) = store();
        let blob_a = put_blob(&s, b"original");
        let base_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_a)]);
        let base_commit = make_commit(&s, base_tree, &[], "initial");
        let target_tree = make_tree(&s, vec![]);
        let target_commit = make_commit(&s, target_tree, &[base_commit], "remove a.txt");
        let blob_modified = put_blob(&s, b"modified content");
        let ours_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_modified)]);
        let r = cherry_pick(&s, target_commit, ours_tree, None).unwrap();
        assert!(r.has_conflicts());
        assert_eq!(r.conflicts[0].kind, ConflictKind::DeleteModify);
        assert_eq!(r.conflicts[0].path, "a.txt");
    }

    #[test]
    fn adds_multiple_files() {
        let (_d, s) = store();
        let blob_a = put_blob(&s, b"aaa");
        let base_tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_a)]);
        let base_commit = make_commit(&s, base_tree, &[], "initial");
        let blob_b = put_blob(&s, b"bbb");
        let blob_c = put_blob(&s, b"ccc");
        let blob_d = put_blob(&s, b"ddd");
        let target_tree = make_tree(
            &s,
            vec![
                entry(b"a.txt", EntryMode::Blob, blob_a),
                entry(b"b.txt", EntryMode::Blob, blob_b),
                entry(b"c.txt", EntryMode::Blob, blob_c),
                entry(b"d.txt", EntryMode::Blob, blob_d),
            ],
        );
        let target_commit = make_commit(&s, target_tree, &[base_commit], "add b, c, d");
        let r = cherry_pick(&s, target_commit, base_tree, None).unwrap();
        assert!(!r.has_conflicts());
        assert_eq!(tree_entries(&s, r.tree_hash).len(), 4);
    }

    #[test]
    fn non_commit_input_returns_error() {
        let (_d, s) = store();
        let blob_hash = put_blob(&s, b"just a blob");
        let empty_tree = make_tree(&s, vec![]);
        let err = cherry_pick(&s, blob_hash, empty_tree, None).unwrap_err();
        assert!(matches!(err, CherryPickError::NotACommit));
    }

    #[test]
    fn root_commit_onto_empty_ours() {
        let (_d, s) = store();
        let blob_a = put_blob(&s, b"aaa");
        let blob_b = put_blob(&s, b"bbb");
        let root_tree = make_tree(
            &s,
            vec![
                entry(b"a.txt", EntryMode::Blob, blob_a),
                entry(b"b.txt", EntryMode::Blob, blob_b),
            ],
        );
        let root_commit = make_commit(&s, root_tree, &[], "root");
        let empty_tree = make_tree(&s, vec![]);
        let r = cherry_pick(&s, root_commit, empty_tree, None).unwrap();
        assert!(!r.has_conflicts());
        assert_eq!(tree_entries(&s, r.tree_hash).len(), 2);
    }

    /// Build a 2-parent merge commit. Parent 1 has `a.txt`, parent 2 has
    /// `b.txt`; the merge tree has both, plus `m.txt` unique to the merge.
    fn merge_fixture(s: &ObjectStore) -> (Hash, Hash, Hash) {
        let a = put_blob(s, b"a");
        let b = put_blob(s, b"b");
        let p1_tree = make_tree(s, vec![entry(b"a.txt", EntryMode::Blob, a)]);
        let p2_tree = make_tree(s, vec![entry(b"b.txt", EntryMode::Blob, b)]);
        let p1 = make_commit(s, p1_tree, &[], "p1");
        let p2 = make_commit(s, p2_tree, &[], "p2");
        let m = put_blob(s, b"m");
        let merge_tree = make_tree(
            s,
            vec![
                entry(b"a.txt", EntryMode::Blob, a),
                entry(b"b.txt", EntryMode::Blob, b),
                entry(b"m.txt", EntryMode::Blob, m),
            ],
        );
        let merge_commit = make_commit(s, merge_tree, &[p1, p2], "merge");
        (merge_commit, p1, p2)
    }

    #[test]
    fn merge_without_mainline_is_rejected() {
        let (_d, s) = store();
        let (merge_commit, _p1, _p2) = merge_fixture(&s);
        let empty = make_tree(&s, vec![]);
        let err = cherry_pick(&s, merge_commit, empty, None).unwrap_err();
        assert!(matches!(err, CherryPickError::MergeNeedsMainline));
    }

    #[test]
    fn mainline_on_non_merge_is_rejected() {
        let (_d, s) = store();
        let blob_a = put_blob(&s, b"aaa");
        let tree = make_tree(&s, vec![entry(b"a.txt", EntryMode::Blob, blob_a)]);
        let root = make_commit(&s, tree, &[], "root");
        let child = make_commit(&s, tree, &[root], "child");
        let empty = make_tree(&s, vec![]);
        let err = cherry_pick(&s, child, empty, Some(1)).unwrap_err();
        assert!(matches!(err, CherryPickError::MainlineForNonMerge));
    }

    #[test]
    fn out_of_range_mainline_is_rejected() {
        let (_d, s) = store();
        let (merge_commit, _p1, _p2) = merge_fixture(&s);
        let empty = make_tree(&s, vec![]);
        let err = cherry_pick(&s, merge_commit, empty, Some(3)).unwrap_err();
        assert!(matches!(
            err,
            CherryPickError::BadMainline {
                given: 3,
                parents: 2
            }
        ));
    }

    #[test]
    fn mainline_selects_the_chosen_parent_as_base() {
        // Picking the merge onto an empty tree with `-m 1` replays the merge's
        // diff against parent 1 (which has a.txt) → adds b.txt + m.txt; with
        // `-m 2` (parent has b.txt) → adds a.txt + m.txt. Different bases ⇒
        // different results, proving the mainline actually selects the parent.
        let (_d, s) = store();
        let (merge_commit, _p1, _p2) = merge_fixture(&s);
        let empty = make_tree(&s, vec![]);
        let r1 = cherry_pick(&s, merge_commit, empty, Some(1)).unwrap();
        let r2 = cherry_pick(&s, merge_commit, empty, Some(2)).unwrap();
        assert_ne!(
            r1.tree_hash, r2.tree_hash,
            "different mainline parents must yield different patches"
        );
    }
}
